### PR TITLE
cmd/{containerboot,k8s-operator},k8s-operator,kube: add ProxyGroup controller

### PR DESCRIPTION
**Pull request by tomhjp**
Implements the controller for the new ProxyGroup CRD, designed for running proxies in a high availability configuration. Each proxy gets its own config and state Secret, and its own tailscale node ID.

We are currently mounting all of the config secrets into the container, but will stop mounting them and instead read them 
directly from the kube API once #13578 is implemented.

Updates #13406

### CODE REVIEW

File: cmd/k8s-operator/proxygroup.go [hunk 1]
```diff
@@ -0,0 +1,509 @@
...
+			mak.Set(&cfgSecret.StringData, tsoperator.TailscaledConfigFileName(cap), string(cfgJSON))
+		}
+
+		if existingCfgSecret != nil {
+			logger.Debugf("patching the existing ProxyGroup config Secret %s with tailscaled config %s", cfgSecret.Name, sanitizeConfigBytes(latestCfg))
+			if err := r.Patch(ctx, cfgSecret, client.MergeFrom(existingCfgSecret)); err != nil {
+				return "", err
+			}
+		} else {
+			logger.Debugf("creating a new Secret %s for the ProxyGroup with tailscaled config %s", cfgSecret.Name, sanitizeConfigBytes(latestCfg))
```
[the remark this answers is not in the document]
Comment (tomhjp) [reply cmd/k8s-operator/proxygroup.go:382]: lol, nice fix 👌 🚢

File: cmd/k8s-operator/operator.go [hunk 2]
```diff
@@ -646,6 +669,27 @@ func proxyClassHandlerForConnector(cl client.Client, logger *zap.SugaredLogger)
 	}
 }
 
+// proxyClassHandlerForConnector returns a handler that, for a given ProxyClass,
+// returns a list of reconcile requests for all Connectors that have
+// .spec.proxyClass set.
+func proxyClassHandlerForProxyGroup(cl client.Client, logger *zap.SugaredLogger) handler.MapFunc {
```
Comment (irbekrm) [inline cmd/k8s-operator/operator.go:675]: We will need also need to reconcile any ProxyGroups to whom a default ProxyClass applies when that ProxyClass changes. We can make that change in a separate PR though- I think some other reconcilers might have the same issue.
Perhaps you could add a TODO here

File: cmd/k8s-operator/proxygroup.go [hunk 3]
```diff
@@ -0,0 +1,509 @@
...
+		return setStatusReady(pg, metav1.ConditionFalse, reasonProxyGroupCreating, message)
+	}
+
+	if len(pg.Status.Devices) > desiredReplicas {
+		message := fmt.Sprintf("waiting for %d ProxyGroup pods to shut down", len(pg.Status.Devices)-desiredReplicas)
+		logger.Debug(message)
+		return setStatusReady(pg, metav1.ConditionFalse, reasonProxyGroupCreating, message)
+	}
+
+	logger.Info("ProxyGroup resources synced")
```
[the remark this answers is not in the document]
Comment (irbekrm) [reply cmd/k8s-operator/proxygroup.go:157]: Perhaps you could add a TODO so we don't forget

File: cmd/k8s-operator/proxygroup.go [hunk 4]
```diff
@@ -0,0 +1,509 @@
...
+
+func (r *ProxyGroupReconciler) maybeProvision(ctx context.Context, pg *tsapi.ProxyGroup) error {
+	logger := r.logger(pg.Name)
+	r.mu.Lock()
+	r.proxyGroups.Add(pg.UID)
+	gaugeProxyGroupResources.Set(int64(r.proxyGroups.Len()))
+	r.mu.Unlock()
+
+	var proxyClass *tsapi.ProxyClass
+	if pg.Spec.ProxyClass != "" {
```
Comment (irbekrm) [inline cmd/k8s-operator/proxygroup.go:169]: Or default?

Comment (irbekrm) [reply cmd/k8s-operator/proxygroup.go:169]: See https://github.com/tailscale/tailscale/blob/9bd158cc09d926c4dd6a9311c3fd68a4ed01a6b1/cmd/k8s-operator/svc.go#L67

Comment (irbekrm) [reply cmd/k8s-operator/proxygroup.go:169]: https://github.com/tailscale/tailscale/blob/9bd158cc09d926c4dd6a9311c3fd68a4ed01a6b1/cmd/k8s-operator/svc.go#L214-L215

File: cmd/k8s-operator/proxygroup.go [hunk 5]
```diff
@@ -0,0 +1,509 @@
...
+}
+
+func pgTailscaledConfig(pg *tsapi.ProxyGroup, class *tsapi.ProxyClass, idx int32, authKey string, oldSecret *corev1.Secret) (tailscaledConfigs, error) {
+	conf := &ipn.ConfigVAlpha{
+		Version:             "alpha0",
+		AcceptDNS:           "false",
+		AcceptRoutes:        "false", // AcceptRoutes defaults to true
+		Locked:              "false",
+		Hostname:            ptr.To(fmt.Sprintf("%s-%d", pg.Name, idx)),
+		NoStatefulFiltering: "false",
```
Comment (irbekrm) [inline cmd/k8s-operator/proxygroup.go:408]: I think we don't need this. (We should also remove it for the Services)

Comment (tomhjp) [reply cmd/k8s-operator/proxygroup.go]: Cool 👍 just to confirm, I also removed the logic below that flipped it to true for egress proxies

File: cmd/k8s-operator/proxygroup.go (hunk 4 above)
Comment (tomhjp) [reply cmd/k8s-operator/proxygroup.go:169]: As discussed offline I'll defer on this for now

File: cmd/k8s-operator/operator.go (hunk 2 above)
Comment (tomhjp) [reply cmd/k8s-operator/operator.go:718]: I think this one is also covered by the offline discussion to defer using the default?

File: cmd/k8s-operator/proxygroup_specs.go [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+							},
+							Args: []string{
+								"sysctl -w net.ipv4.ip_forward=1 && if sysctl net.ipv6.conf.all.forwarding; then sysctl -w net.ipv6.conf.all.forwarding=1; fi",
+							},
+						},
+					},
+					Containers: []corev1.Container{
+						{
+							Name:  "tailscale",
+							Image: fmt.Sprintf("tailscale/tailscale:%s", selfVersionImageTag()),
```
Comment (irbekrm) [inline cmd/k8s-operator/proxygroup_specs.go:67]: (Here and for the init container) - we should respect the proxy image passed to helm install/via env vars.
It won't be that valuable for the initial release, but it would be better if we do it already now rather than on a subsequent release when it would can be breaking

Comment (irbekrm) [reply cmd/k8s-operator/proxygroup_specs.go:67]: Or do you think it can only be configured via env vars?

File: cmd/k8s-operator/operator.go (hunk 2 above)
Comment (irbekrm) [reply cmd/k8s-operator/operator.go:718]: Yes

File: cmd/k8s-operator/proxygroup.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (irbekrm) [reply cmd/k8s-operator/proxygroup.go:157]: However, you might want to update the `.spec.proxyClass` description so it does not mention default ProxyClass

File: cmd/k8s-operator/proxygroup.go [hunk 7]
```diff
@@ -0,0 +1,509 @@
...
+	if pg.Spec.Type == tsapi.ProxyClassTypeEgress {
+		conf.NoStatefulFiltering = "true"
+	}
+	if shouldAcceptRoutes(class) {
+		conf.AcceptRoutes = "true"
+	}
+
+	if authKey != "" {
+		conf.AuthKey = &authKey
+	} else if shouldRetainAuthKey(oldSecret) {
```
Comment (irbekrm) [inline cmd/k8s-operator/proxygroup.go:420]: This results in the auth key never being cleared out because the `shouldRetainAuthKey` looks for `device_id` field that does not get set to this Secret

Comment (irbekrm) [reply cmd/k8s-operator/proxygroup.go:420]: We should probably retrieve the state Secret and read the field from there

File: cmd/k8s-operator/proxygroup_specs.go [hunk 8]
```diff
@@ -0,0 +1,242 @@
...
+							SecurityContext: &corev1.SecurityContext{
+								Privileged: ptr.To(true),
+							},
+							Command: []string{
+								"/bin/sh",
+								"-c",
+							},
+							Args: []string{
+								"sysctl -w net.ipv4.ip_forward=1 && if sysctl net.ipv6.conf.all.forwarding; then sysctl -w net.ipv6.conf.all.forwarding=1; fi",
```
Comment (irbekrm) [inline cmd/k8s-operator/proxygroup_specs.go:60]: It would have been better to use the existing templates that we have https://github.com/tailscale/tailscale/blob/main/cmd/k8s-operator/deploy/manifests/proxy.yaml
These proxies are meant to be configured in the same way as the ones from the templates- now we'll have the same config in two places (potential for drift, need to make changes in two places etc)

File: cmd/k8s-operator/proxygroup.go [hunk 9]
```diff
@@ -0,0 +1,509 @@
...
+	reasonProxyGroupCreationFailed = "ProxyGroupCreationFailed"
+	reasonProxyGroupCreated        = "ProxyGroupCreated"
+	reasonProxyGroupCreating       = "ProxyGroupCreating"
+	reasonProxyGroupInvalid        = "ProxyGroupInvalid"
+)
+
+var gaugeProxyGroupResources = clientmetric.NewGauge(kubetypes.MetricProxyGroupCount)
+
+// ProxyGroupReconciler syncs ProxyGroup statefulsets with their definition in
+// ProxyGroup CRs.
```
Comment (irbekrm) [inline cmd/k8s-operator/proxygroup.go:53]: Nit:
```suggestion
// ProxyGroupReconciler ensures cluster resources for a ProxyGroup definition.
```
or something. In any case, it's not just StatefulSets

File: cmd/k8s-operator/proxygroup_specs.go [hunk 10]
```diff
@@ -0,0 +1,242 @@
...
+				OwnerReferences: pgOwnerReference(pg),
+			},
+		})
+	}
+
+	return secrets
+}
+
+func secretLabels(app, instance, typ string) map[string]string {
+	return labels(app, instance, map[string]string{
```
Comment (irbekrm) [inline cmd/k8s-operator/proxygroup_specs.go:191]: I think we should try to make this more similar to how the existing proxies work:
- label them with `tailscale.com` labels
- use `tailscale.com` labels for lookup
This is for consistency with the existing proxies, because those labels are more semantically meaningful and because with those we can be most certain that there won't be any surprises (users also created a `Secret` with that label for a different purpose and we are now picking that up

Comment (irbekrm) [reply cmd/k8s-operator/proxygroup_specs.go:191]: For resources created for `ProxyGroup` there would be good to have:
- `tailscale.com/managed: true` - we set that for all other proxy resources
- `tailscale.com/proxy-group: <proxy-group-name>` - for lookup

File: cmd/k8s-operator/proxygroup.go (hunk 4 above)
Comment (irbekrm) [reply cmd/k8s-operator/proxygroup.go:169]: However, you might want to update the `.spec.proxyClass` field description as currently it suggests that the default `ProxyClass` will be used

File: cmd/k8s-operator/proxygroup.go [hunk 11]
```diff
@@ -0,0 +1,509 @@
...
+			logger.Debugf("deleted state Secret %s but got not found", m.stateSecret.Name)
+		}
+		configSecret := m.stateSecret.DeepCopy()
+		configSecret.Name += "-config"
+		if err := r.Delete(ctx, configSecret); err != nil {
+			if !apierrors.IsNotFound(err) {
+				return fmt.Errorf("error deleting config Secret %s: %w", configSecret.Name, err)
+			}
+
+			logger.Debugf("deleted config Secret %s but got not found", configSecret.Name)
```
Comment (irbekrm) [inline cmd/k8s-operator/proxygroup.go:280]: Is this line meant to stay?

File: cmd/k8s-operator/proxygroup.go [hunk 12]
```diff
@@ -0,0 +1,509 @@
...
+		if err != nil {
+			return nil, fmt.Errorf("error retrieving auth key from Secret: %w", err)
+		}
+		conf.AuthKey = key
+	}
+	capVerConfigs := make(map[tailcfg.CapabilityVersion]ipn.ConfigVAlpha)
+	capVerConfigs[95] = *conf
+	// legacy config should not contain NoStatefulFiltering field.
+	conf.NoStatefulFiltering.Clear()
+	capVerConfigs[94] = *conf
```
Comment (irbekrm) [inline cmd/k8s-operator/proxygroup.go:431]: These lines are not needed, this is not assumed to work with proxies older than cap 95

File: cmd/k8s-operator/proxygroup.go [hunk 13]
```diff
@@ -0,0 +1,509 @@
...
+		conf.AuthKey = &authKey
+	} else if shouldRetainAuthKey(oldSecret) {
+		key, err := authKeyFromSecret(oldSecret)
+		if err != nil {
+			return nil, fmt.Errorf("error retrieving auth key from Secret: %w", err)
+		}
+		conf.AuthKey = key
+	}
+	capVerConfigs := make(map[tailcfg.CapabilityVersion]ipn.ConfigVAlpha)
+	capVerConfigs[95] = *conf
```
Comment (irbekrm) [inline cmd/k8s-operator/proxygroup.go:428]: Let's set this to the latest cap, so folks don't attempt to use older proxy images

File: cmd/k8s-operator/proxygroup.go [hunk 14]
```diff
@@ -0,0 +1,509 @@
...
+}
+
+// getNodeMetadata gets metadata for all the pods owned by this ProxyGroup by
+// querying their state Secrets. It may not return the same number of items as
+// specified in the ProxyGroup spec if e.g. it is getting scaled up or down, or
+// some pods have failed to write state.
+func (r *ProxyGroupReconciler) getNodeMetadata(ctx context.Context, pg *tsapi.ProxyGroup) (metadata []nodeMetadata, _ error) {
+	// List all state secrets owned by this ProxyGroup.
+	secrets := &corev1.SecretList{}
+	if err := r.List(ctx, secrets, client.InNamespace(r.tsNamespace), client.MatchingLabels(secretLabels("proxygroup", pg.Name, "state"))); err != nil {
```
Comment (irbekrm) [inline cmd/k8s-operator/proxygroup.go:446]: We should use a `tailscale.com` label as identifier

File: cmd/k8s-operator/proxygroup_specs.go (hunk 10 above)
Comment (irbekrm) [reply cmd/k8s-operator/proxygroup_specs.go:191]: Also would be nice to have a separate `proxyGroupChildResourceLabels` function that takes a `ProxyGroup` so it's clearer exactly what labels all the resources created for a `ProxyGroup` will have and what the values are for a specific `ProxyGroup` will be

Comment (irbekrm) [reply cmd/k8s-operator/proxygroup_specs.go:191]: Last bit - managed labels need to be added to https://github.com/tailscale/tailscale/blob/main/cmd/k8s-operator/sts.go#L101

File: cmd/k8s-operator/proxygroup_specs.go (hunk 8 above)
Comment (tomhjp) [reply cmd/k8s-operator/proxygroup_specs.go:60]: Sorry if I misunderstood, I thought this was part of the clean break we discussed. That said, I would probably prefer to keep it this way so that the specs are defined somewhat completely and all together in the same file, as there's a non-trivial amount of linkage between them all.

My thinking with maintenance is that we won't need to make updates to the original proxies for long anyway, as we'd like to standardise on one workflow (i.e. default ProxyGroup) as soon as possible, but maybe I'm too optimistic about the timeline for that.

File: cmd/k8s-operator/proxygroup.go (hunk 7 above)
Comment (tomhjp) [reply cmd/k8s-operator/proxygroup.go:420]: Oh nice spot, thanks. Will fix that up.

File: cmd/k8s-operator/proxygroup_specs.go (hunk 6 above)
Comment (tomhjp) [reply cmd/k8s-operator/proxygroup_specs.go:67]: Agreed, thanks for catching this, I forgot to consider the env var.

File: cmd/k8s-operator/proxygroup.go [hunk 15]
```diff
@@ -0,0 +1,509 @@
...
+		return fmt.Errorf("error creating ServiceAccount: %w", err)
+	}
+	role := pgRole(pg, r.tsNamespace)
+	if _, err := createOrUpdate(ctx, r.Client, r.tsNamespace, role, func(r *rbacv1.Role) {
+		r.ObjectMeta.Labels = role.ObjectMeta.Labels
+		r.ObjectMeta.Annotations = role.ObjectMeta.Annotations
+		r.ObjectMeta.OwnerReferences = role.ObjectMeta.OwnerReferences
+		r.Rules = role.Rules
+	}); err != nil {
+		return fmt.Errorf("error creating Role: %w", err)
```
Comment (irbekrm) [inline cmd/k8s-operator/proxygroup.go:209]: (Here and in other places in this file):
```
return fmt.Errorf("error ensuring Role: %w", err)
```
or something along those lines, to make it less confusing if it is a subsequent update that fails

File: cmd/k8s-operator/proxygroup_specs.go (hunk 6 above)
Comment (tomhjp) [reply cmd/k8s-operator/proxygroup_specs.go]: Fixed in e1d2b459b19e741d052aca13d3119f5e00960491

File: cmd/k8s-operator/proxygroup_specs.go (hunk 10 above)
Comment (tomhjp) [reply cmd/k8s-operator/proxygroup_specs.go]: 👍 thanks for the notes, added in 101bd89efdab7da6f6fb0488c6ced66c0faf68b7

**Comment by tomhjp** [comment]
Thanks for the review! I think this is ready for another pass @irbekrm

File: cmd/k8s-operator/proxygroup_test.go [hunk 16]
```diff
@@ -0,0 +1,207 @@
...
+
+		tsoperator.SetProxyGroupCondition(pg, tsapi.ProxyGroupReady, metav1.ConditionFalse, reasonProxyGroupCreating, "0/2 ProxyGroup pods running", 0, cl, zl.Sugar())
+		expectEqual(t, fc, pg, nil)
+		if expected := 1; reconciler.proxyGroups.Len() != expected {
+			t.Fatalf("expected %d recorders, got %d", expected, reconciler.proxyGroups.Len())
+		}
+		expectProxyGroupResources(t, fc, pg, true)
+	})
+
+	t.Run("create state secrets with fake node info, and see metadata appear in status", func(t *testing.T) {
```
Comment (irbekrm) [inline cmd/k8s-operator/proxygroup_test.go:69]: I think we generally prefer to have the test case etc names with underscores instead of spaces for easier lookup if it fails

**Review by irbekrm — CHANGES_REQUESTED** [review]
Thanks for making the changes @tomhjp, I've tested this with an egress ConfigMap on top + the changes from this code review + a couple more small changes that I will make and exposing/modifying/unexposing an egress service works as expected 🥳 

Apart from these few things, this looks ready to go in

File: cmd/k8s-operator/proxygroup_specs.go [hunk 17]
```diff
@@ -0,0 +1,242 @@
...
+			Value: "kube:$(POD_NAME)",
+		},
+		{
+			Name:  "TS_EXPERIMENTAL_VERSIONED_CONFIG_DIR",
+			Value: "/etc/tsconfig/$(POD_NAME)",
+		},
+	}
+
+	return envs
+}
```
[the remark this answers is not in the document]
Comment (irbekrm) [reply cmd/k8s-operator/proxygroup_specs.go:241]: With that, I think we'll also need to pull in the default proxy class, else that helm values block becomes too inconsistent. But not in this PR I think

[the remark this answers is not in the document]
Comment (tomhjp) [reply cmd/k8s-operator/proxygroup_specs.go]: Yeah agreed about using the default proxy class - I also noticed while I was in the area that I had failed to thread through the default tags from the helm chart, so I fixed that in this PR, but can follow up on the default proxy class in another PR.

File: cmd/k8s-operator/proxygroup_specs.go (hunk 8 above)
Comment (tomhjp) [reply cmd/k8s-operator/proxygroup_specs.go:58]: That's fair, it is definitely a lot noisier visually to read the Go version than the yaml version. As discussed offline I'll note this down as a follow-up to save adding too much churn in this PR.

**Comment by tomhjp** [comment]
Thanks for testing the full flow, and catching some bugs! I think it should be fixed up now other than my two follow-ups to convert to using proxy.yaml as the skeleton and honour the default ProxyClass setting.

File: cmd/k8s-operator/proxygroup_specs.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (irbekrm) [reply cmd/k8s-operator/proxygroup_specs.go]: >  I also noticed while I was in the area that I had failed to thread through the default tags from the helm chart, so I fixed that in this PR

Oh, that's great, I didn't spot that

> can follow up on the default proxy class in another PR

another PR sounds good 👍🏼

**Review by irbekrm — APPROVED** [review]
Thanks a lot for getting this one into ready state!

Looks good to me 🥳

**Comment by tomhjp** [comment]
Thanks for the quick reviews! 🚀
